### PR TITLE
Fix #4605 - Do not cache commands with Contains on Enumerable parameter.

### LINKED ISSUE #4605: Query: Revisit Enumerable.Contains command caching
Different SQL is generated for Enumerable.Contains depending on the contents of the enumerable. The second level (command) caching of these commands does not currently have any kind of eviction policy and may grow out of control.

Possible solutions to the issue include:
* Introduce an eviction policy for the command cache
* Only add commands to the cache if the same cache key is encountered more than once
* Do not enter Enumerable.Contains commands into the cache
* Parameterize the Enumerable (TVP) in the query
* Support partial command construction / SQL generation

### PR DESCRIPTION
**Pull request by mikary**
May need to revisit the `CommandCacheKey` comparison after  #4575 is merged
@anpete

### CODE REVIEW

**Comment by anpete** [comment]
Can you add tests?

File: src/Microsoft.EntityFrameworkCore.Relational/Query/Sql/IQuerySqlGenerator.cs [hunk 1]
```diff
@@ -12,6 +12,8 @@ public interface IQuerySqlGenerator
     {
         IRelationalCommand GenerateSql([NotNull] IReadOnlyDictionary<string, object> parameterValues);
 
+        bool CanCacheCommand { get; }
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Sql/IQuerySqlGenerator.cs]: IsCacheable?

**Comment by mikary** [comment]
Updated based on feedback / discussion
Added a test to cover the Contains on Enumerable parameter in a subquery case, though this may not be necessary now that the IsCacheable state is computed after SQL gen.

File: src/Microsoft.EntityFrameworkCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs [hunk 2]
```diff
@@ -28,6 +28,7 @@ public class DefaultQuerySqlGenerator : ThrowingExpressionVisitor, ISqlExpressio
         private IRelationalCommandBuilder _relationalCommandBuilder;
         private IReadOnlyDictionary<string, object> _parametersValues;
         private ParameterNameGenerator _parameterNameGenerator;
+        private bool _hasParameterLiterals;
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: Can we remove this and use IsCacheable directly?

File: src/Microsoft.EntityFrameworkCore.Relational/Query/Internal/ShaperCommandContext.cs [hunk 3]
```diff
@@ -82,9 +81,22 @@ public override bool Equals(object obj)
...
+            IRelationalCommand relationalCommand;
+            var key = new CommandCacheKey(parameters);
+
+            if (!_commandCache.TryGetValue(key, out relationalCommand))
+            {
+                var generator = QuerySqlGeneratorFactory();
+
+                relationalCommand = generator.GenerateSql(parameters);
+
+                if(generator.IsCacheable)
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Internal/ShaperCommandContext.cs]: Formatting. Consider R# code cleanup.

**Comment by anpete** [comment]
:shipit:

**Pull request merged**